File: Remove_Nth_Node_From_End_of_List.cpp
/*
Time Complexity: O(n)To go through the whole linked list) 
Space Complexity: O(1)
Did this code successfully run on Leetcode :yes
Any problem you faced while coding this : no
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* temp = new ListNode(-1);
        temp->next = head;
        ListNode* right = temp;
        ListNode* left = temp;
        for(int i=0;i<n;i++){
            right=right->next;
        }
        while(right->next!=NULL){
            left=left->next;
            right=right->next;
        }
        left->next=left->next->next;
        return temp->next;
    }
};

File: remove-nth-node-from-end-of-list.js
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No, I've done this problem before

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    if (head.next == null && n >= 1) return null; 
    let distanceFromPos = 0;
    let nthNode = head;
    let tail = head;
    while (tail.next != null) {
        if (distanceFromPos != n) {
            distanceFromPos++;
        } else {
            nthNode = nthNode.next;
        }
        tail = tail.next;
    }
    
    if (distanceFromPos != n && nthNode == head && n > 1) {
        head = head.next
    } else {
        nthNode.next = nthNode.next.next;
    }
    return head;
};
